design/deploy.md

```diff
@@ -0,0 +1,98 @@
+  - Additional type-specific configuration (examples below).
+  - We could also treat environments as a special case of the existing build concept, and include them in the `build` section.
+
+- Add a `deploy` command.
+  - Pushes to the hosting service for the current environment.
+  - Under the `default` environment, offers to help set up a deploy environment.
+     - Could have an `init`-style menu to choose from the available types.
+  - With the `--open` flag, also opens a browser (same as `serve`).
+  - Print the commands we're running so that users can do it themselves.
+
+## Environments
+
+### App Engine
+
+- Has a [free tier](https://cloud.google.com/free/).
+
+- Supports Node.js flex environment.
+
+- Custom `polymer.json` environment configuration:
+  - **project_id**: The App Engine project ID to deploy to.
+
+- `build`
+
+  - Runs both a bundled and unbundled standard build.
+
+  - `app.yaml`
+     - Mostly just specifies the `nodejs` runtime.
+
+  - `package.json`
+     - Note: this can't be the same as the existing *project* `package.json`.  This config defines a specific AppEngine deployment. It has different dependencies to the project, sets the deployed node version, etc.
+
+  - `server.js`
+     - An express HTTP server.
+     - Not polyserve, because it is a bit heavy with development-oriented features. Prefer something as simple as possible since we'd like users to understand it completely.
+     - Implementation is stamped out from a Yeoman template (as opposed to e.g. a binary dependency and generated config file) so that it's easy to customize later.
+     - Serves the main file from all URLs.
+     - Serves bundled or unbundled depending on client capabilities.
+     - Supports server push manifest.
+         - Could we factor out polyserve's push manifest support as an express middleware in a separate repo?
+     - Note: App Engine also supports a built-in static file server configurable directly in `app.yaml`. This is what the [Polymer App Toolbox tutorial](https://www.polymer-project.org/2.0/start/toolbox/deploy#deploy-with-appengine) demonstrates. It would support serving the main file from all URLs, but it not Server Push or User-Agent switching behavior. To show off PRPL we need to include a custom server.
+
+- `serve`
+  - Runs `build`, `npm install`, and `npm start`.
+     - App Engine will give us the Node.js runtime version configured in `package.json` `engines`, but it might not match what the user has installed locally, so behavior could theoretically differ. `npm` should warn if they mismatch, which is probably good enough. The `--engine-strict` flag would make this a hard error.
+     - Note: `dev_appserver.py` does not support the nodejs flex runtime. This wouldn't buy us anything anyway, since flex environments are docker containers with standard runtimes.
+
+- `deploy`
+  - Runs `build`, `npm install`, and `gcloud app deploy --project="$PROJECT_ID"`.
```

Comment:
It's interesting that the deploy command runs the build command. Like we talked about yesterday, this is a new dependency between commands. We should talk about this outside of the App Engine environment and show how an environment would invoke another command. We should also talk about whether it's ok for deploy to modify `build` output, or whether new artifacts need to happen in `build`, or if `deploy` gets it's own file system space to transform the build output.